deluge/decorators.py

```diff
@@ -159,3 +161,36 @@ def depr_func(*args, **kwargs):
         return func(*args, **kwargs)
 
     return depr_func
+
+
+class CoroDeferred(defer.Deferred):
+    """Deferred that will dynamically act like a coroutine instead of a Deferred if it's awaited."""
+
+    def __init__(self, coro):
+        super().__init__()
+        self.coro = coro
+        self.awaited = None
+        self.activate_deferred = reactor.callLater(0, self.activate)
+
+    def __await__(self):
+        if self.awaited in [None, True]:
+            self.awaited = True
+            return self.coro.__await__()
+        return self.__iter__()
+
+    def activate(self):
+        """If the result wasn't awaited before the next context switch, we turn it into a deferred."""
+        if self.awaited is None:
+            self.awaited = False
+            d = defer.Deferred.fromCoroutine(self.coro)
+            d.chainDeferred(self)
+
+
+def ensure_deferred(f):
```

Comment:
Twisted already has [maybeDeferred](https://twistedmatrix.com/documents/current/api/twisted.internet.defer.html#maybeDeferred) so it might be confusing to name the decorator `maybe_deferred` so what if we went the other way with `maybe_coroutine`?